src/Compilers/Core/Portable/InternalUtilities/FatalError.cs

```diff
@@ -74,6 +75,13 @@ public static void OverwriteHandler(Action<Exception>? value)
             s_fatalHandler = value;
         }
 
+        private static bool IsCurrentOperationBeingCancelled(Exception exception, CancellationToken cancellationToken)
+#if NET20
+            => exception is OperationCanceledException cancellationException;
+#else
+            => exception is OperationCanceledException cancellationException && cancellationException.CancellationToken == cancellationToken;
```

Comment:
❗️ The cancellation tokens do not need to match. The only requirement is that `cancellationToken` has requested cancellation.

Reply:
@jasonmalinowski thinks otherwise. Seems like we need to get to an agreement what the right approach is.

Reply:
> This would disallow things like linked cancellation tokens.

My belief is that in the linked case you potentially need to catch and rethrow with the original cancellation token because others still assume that the cancellation tokens must match, i.e. don't leak any linked tokens outside your function if you internally created one. It's all a question of what your _caller_ expects, even if the TPL now behaves differently.

What also worries me still is these checks in the TPL -- clearly *something* is making decisions on the underlying tokens, or is this dead code?

https://github.com/dotnet/runtime/blob/26c11e02a6e4635396f7921c5f5465370b37d1b2/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs#L2392
https://github.com/microsoft/referencesource/blob/17b97365645da62cf8a49444d979f94a59bbb155/mscorlib/system/threading/Tasks/Parallel.cs#L3544-L3572

Reply:
(if you search our internal index on references to OperationCanceledException.CancellationToken there are a lot of other places also checking, both in Roslyn and in places that may call Roslyn. So if we're no longer maintaining that for Roslyn APIs then I'm a bit worried what we might break in the process.

Reply:
And if there was a memo that the intended pattern has changed for .NET I just missed the memo. But I'm not sure then how different patterns are supposed to combine. If a caller is expecting the stricter requirement that tokens match, then it would seem we should still be enforcing in our internal checks so we don't leak an exception that breaks those assumptions...

Reply:
👍

Reply:
Let me change this to `cancellationToken.IsCancellationRequested` for now and then we can strengthen the condition in a follow up, once we agree on it.